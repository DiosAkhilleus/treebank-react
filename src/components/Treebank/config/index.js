import {
  getConfig,
  getColor,
} from './config';

export {
  getConfig,
  getColor,
};
