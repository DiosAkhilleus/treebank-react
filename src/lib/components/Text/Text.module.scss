.text {
  background-color: rgb(242, 242, 242);
  text-align: left;

  padding-top: 0.2rem;
  padding-bottom: 0.2rem;
  padding-left: 0.6rem;
  padding-right: 0.6rem;

  margin-bottom: 0.2rem;
}

.word {
  cursor: pointer;
  padding-left: 1px;
  padding-right: 1px;
  font-weight: 450;

  &:hover {
    background-color: #5BC8DC;
  }
}

.active {
  background-color: rgb(246, 217, 24);
}

.elliptic {
  font-size: 0.7rem;
}
