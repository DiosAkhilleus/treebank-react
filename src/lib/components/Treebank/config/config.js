import aldtMiscGrc from './arethusa-configs/aldt-misc-grc.json';
import aldtMisc from './arethusa-configs/aldt-misc.json';
import aldt2grcStage from './arethusa-configs/aldt2grc-stage.json';
import aldt2grc from './arethusa-configs/aldt2grc.json';
import aldt2latStage from './arethusa-configs/aldt2lat-stage.json';
import aldt2lat from './arethusa-configs/aldt2lat.json';
import elmlatch1 from './arethusa-configs/elmlatch1.json';
import elmlatch10 from './arethusa-configs/elmlatch10.json';
import elmlatch10b from './arethusa-configs/elmlatch10b.json';
import elmlatch11 from './arethusa-configs/elmlatch11.json';
import elmlatch2 from './arethusa-configs/elmlatch2.json';
import elmlatch3 from './arethusa-configs/elmlatch3.json';
import elmlatch4 from './arethusa-configs/elmlatch4.json';
import elmlatch5 from './arethusa-configs/elmlatch5.json';
import elmlatch6 from './arethusa-configs/elmlatch6.json';
import elmlatch7 from './arethusa-configs/elmlatch7.json';
import elmlatch9 from './arethusa-configs/elmlatch9.json';
import english from './arethusa-configs/english.json';
import gclat from './arethusa-configs/gclat.json';
import harrington from './arethusa-configs/harrington.json';
import jmhgreek from './arethusa-configs/jmhgreek.json';
import lyonGrec from './arethusa-configs/lyonGrec.json';
import lyonLatin from './arethusa-configs/lyonLatin.json';
import midlat from './arethusa-configs/midlat.json';
import morphgrc from './arethusa-configs/morphgrc.json';
import morphlat from './arethusa-configs/morphlat.json';
import pedalion from './arethusa-configs/pedalion.json';
import persian from './arethusa-configs/persian.json';
import sanskrit from './arethusa-configs/sanskrit.json';
import smyth2 from './arethusa-configs/smyth2.json';
import smyth3 from './arethusa-configs/smyth3.json';
import snap from './arethusa-configs/snap.json';
import testlat from './arethusa-configs/testlat.json';
import turkish from './arethusa-configs/turkish.json';
import ud from './arethusa-configs/ud.json';
import vortex from './arethusa-configs/vortex.json';
import wheelock from './arethusa-configs/wheelock.json';
import wheelock13 from './arethusa-configs/wheelock1-3.json';
import wheelock15 from './arethusa-configs/wheelock15.json';
import wheelock17 from './arethusa-configs/wheelock17.json';
import wheelock22 from './arethusa-configs/wheelock22.json';
import wheelock24 from './arethusa-configs/wheelock24.json';
import wheelock30 from './arethusa-configs/wheelock30.json';
import wheelock46 from './arethusa-configs/wheelock4-6.json';
import whitaker from './arethusa-configs/whitaker.json';

const configMap = {
  'aldt-misc-grc': aldtMiscGrc,
  'aldt-misc': aldtMisc,
  'aldt2grc-stage': aldt2grcStage,
  aldt2grc,
  'aldt2lat-stage': aldt2latStage,
  aldt2lat,
  elmlatch1,
  elmlatch10,
  elmlatch10b,
  elmlatch11,
  elmlatch2,
  elmlatch3,
  elmlatch4,
  elmlatch5,
  elmlatch6,
  elmlatch7,
  elmlatch9,
  english,
  gclat,
  harrington,
  jmhgreek,
  lyonGrec,
  lyonLatin,
  midlat,
  morphgrc,
  morphlat,
  pedalion,
  persian,
  sanskrit,
  smyth2,
  smyth3,
  snap,
  testlat,
  turkish,
  ud,
  vortex,
  wheelock,
  'wheelock1-3': wheelock13,
  wheelock15,
  wheelock17,
  wheelock22,
  wheelock24,
  wheelock30,
  'wheelock4-6': wheelock46,
  whitaker,
};

const getConfig = (format, lang) => {
  // Backwards compatibility
  const key = format === 'aldt' ? `aldt2${lang}` : format;

  if (configMap[key]) {
    return configMap[key];
  }

  if (lang === 'grc') {
    return aldtMiscGrc;
  }

  return aldtMisc;
};

const getColor = (config, postag) => {
  if (postag && config.plugins && config.plugins.morph) {
    const { postagSchema, styledThrough, attributes } = config.plugins.morph;
    const index = postagSchema.findIndex((n) => n === styledThrough);
    const mapping = attributes[styledThrough];

    if (mapping && index > -1) {
      const values = Object.values(mapping.values);
      const valueIndex = values.findIndex(({ postag: abbreviation }) => (
        abbreviation === postag[index]
      ));

      if (valueIndex > -1) {
        const value = values[valueIndex];

        if (value.style && value.style.color) {
          return value.style.color;
        }
      }
    }
  }

  return null;
};

const deconstructPostag = (config, postag) => {
  const deconstructedPostag = [];

  if (postag && config.plugins && config.plugins.morph) {
    const { postagSchema, attributes } = config.plugins.morph;

    postagSchema.forEach((type, index) => {
      const attribute = attributes[type];
      const name = attribute.long || attribute.short || type;
      const values = Object.entries(attribute.values);
      const match = values.find(([, { postag: abbreviation }]) => abbreviation === postag[index]);

      if (match) {
        deconstructedPostag.push([
          name,
          match[1].long || match[1].short || match[0],
        ]);
      }
    });
  }

  return deconstructedPostag;
};

export {
  getConfig,
  getColor,
  deconstructPostag,
};
