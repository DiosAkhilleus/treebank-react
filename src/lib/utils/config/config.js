const configMap = {
  'aldt-misc-grc': true,
  'aldt-misc': true,
  'aldt2grc-stage': true,
  aldt2grc: true,
  'aldt2lat-stage': true,
  aldt2lat: true,
  elmlatch1: true,
  elmlatch10: true,
  elmlatch10b: true,
  elmlatch11: true,
  elmlatch2: true,
  elmlatch3: true,
  elmlatch4: true,
  elmlatch5: true,
  elmlatch6: true,
  elmlatch7: true,
  elmlatch9: true,
  english: true,
  gclat: true,
  harrington: true,
  jmhgreek: true,
  lyonGrec: true,
  lyonLatin: true,
  midlat: true,
  morphgrc: true,
  morphlat: true,
  pedalion: true,
  persian: true,
  sanskrit: true,
  smyth2: true,
  smyth3: true,
  snap: true,
  testlat: true,
  turkish: true,
  ud: true,
  vortex: true,
  wheelock: true,
  'wheelock1-3': true,
  wheelock15: true,
  wheelock17: true,
  wheelock22: true,
  wheelock24: true,
  wheelock30: true,
  'wheelock4-6': true,
  whitaker: true,
};

const getConfig = (format, lang, callback) => {
  // Backwards compatibility
  const key = format === 'aldt' ? `aldt2${lang}` : format;
  const config = configMap[key] ? key : (lang === 'grc' ? 'aldt-misc-grc' : 'aldt-misc')

  import(`./arethusa-configs/${config}.json`).then(callback);
};

const getColor = (config, postag) => {
  if (postag && config.plugins && config.plugins.morph) {
    const { postagSchema, styledThrough, attributes } = config.plugins.morph;
    const index = postagSchema.findIndex((n) => n === styledThrough);
    const mapping = attributes[styledThrough];

    if (mapping && index > -1) {
      const values = Object.values(mapping.values);
      const valueIndex = values.findIndex(({ postag: abbreviation }) => (
        abbreviation === postag[index]
      ));

      if (valueIndex > -1) {
        const value = values[valueIndex];

        if (value.style && value.style.color) {
          return value.style.color;
        }
      }
    }
  }

  return null;
};

const deconstructPostag = (config, postag) => {
  const deconstructedPostag = [];

  if (postag && config.plugins && config.plugins.morph) {
    const { postagSchema, attributes } = config.plugins.morph;

    postagSchema.forEach((type, index) => {
      const attribute = attributes[type];
      const name = attribute.long || attribute.short || type;
      const values = Object.entries(attribute.values);
      const match = values.find(([, { postag: abbreviation }]) => abbreviation === postag[index]);

      if (match) {
        deconstructedPostag.push([
          name,
          match[1].long || match[1].short || match[0],
        ]);
      }
    });
  }

  return deconstructedPostag;
};

export {
  getConfig,
  getColor,
  deconstructPostag,
};
