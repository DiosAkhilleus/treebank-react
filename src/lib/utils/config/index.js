import {
  getConfig,
  getColor,
  deconstructPostag,
} from './config';

export {
  getConfig,
  getColor,
  deconstructPostag,
};
