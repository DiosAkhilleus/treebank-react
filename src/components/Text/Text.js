import React, { Fragment } from 'react';
import { object, func } from 'prop-types';

import { sentenceType, wordType } from '../../types';

import styles from './Text.module.scss';

import { getColor } from '../Treebank/config';

// eslint-disable-next-line react/prop-types
const wordToSpan = (config, active, toggleActive, word) => {
  const { $: { id, form, postag } } = word;
  const color = getColor(config, postag);
  const isActive = active && active.$.id === id;
  const className = isActive ? [styles.word, styles.active].join(' ') : styles.word;

  return (
    <Fragment key={id}>
      <span
        onClick={() => toggleActive(word)}
        className={className}
        style={{ color }}
      >
        {form}
      </span>
      {' '}
    </Fragment>
  );
};

const Text = ({
  sentence, active, toggleActive, config,
}) => (
  <div className={styles.sentence}>
    <p>
      {sentence.word.map((word) => wordToSpan(config, active, toggleActive, word))}
    </p>
  </div>
);

Text.propTypes = {
  sentence: sentenceType.isRequired,
  active: wordType,
  toggleActive: func,
  // eslint-disable-next-line react/forbid-prop-types
  config: object.isRequired,
};

Text.defaultProps = {
  active: null,
  toggleActive: () => {},
};

export default Text;
