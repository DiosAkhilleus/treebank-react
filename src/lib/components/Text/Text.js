import React, { Fragment } from 'react';
import { object, func } from 'prop-types';

import { sentenceType, wordType } from '../../types';

import styles from './Text.module.scss';

import { getColor } from '../../utils/config';

const formatId = (id) => id.padStart(4, '0');

const wordId = ({ $: { id, insertion_id: insertionId } }) => (
  formatId(insertionId || id)
);

const compareWords = (wordA, wordB) => {
  const idA = wordId(wordA);
  const idB = wordId(wordB);

  if (idA < idB) {
    return -1;
  }

  if (idB < idA) {
    return 1;
  }

  return 0;
};

const wordToSpan = (word, config, active, toggleActive) => {
  const {
    $: {
      id, form, postag, artificial,
    },
  } = word;
  const color = getColor(config, postag);
  const classes = [styles.word];

  if (active && active.$.id === id) {
    classes.push(styles.active);
  }

  if (artificial === 'elliptic') {
    classes.push(styles.elliptic);
  }

  const onClick = () => {
    toggleActive(word);
  };
  const onKeyDown = (event) => {
    const { key } = event;

    if (key === 'Enter') {
      onClick();
    }
  };

  return (
    <Fragment key={id}>
      <span
        role="button"
        tabIndex="0"
        onClick={onClick}
        onKeyDown={onKeyDown}
        className={classes.join(' ')}
        style={{ color }}
      >
        {form}
      </span>
      {' '}
    </Fragment>
  );
};

const Text = ({
  sentence, active, toggleActive, config,
}) => {
  const wordsCopy = [...sentence.word];
  const spans = wordsCopy
    .sort(compareWords)
    .map((word) => wordToSpan(word, config, active, toggleActive));

  return (
    <div className={styles.text}>
      <p>
        {spans}
      </p>
    </div>
  );
};

Text.propTypes = {
  sentence: sentenceType.isRequired,
  active: wordType,
  toggleActive: func,
  // eslint-disable-next-line react/forbid-prop-types
  config: object.isRequired,
};

Text.defaultProps = {
  active: null,
  toggleActive: () => {},
};

export default Text;
