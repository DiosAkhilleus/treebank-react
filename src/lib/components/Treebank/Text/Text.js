import React from 'react';

import SentenceContext from '../sentence-context';

import Text from '../../Text';

const TextWithContext = () => (
  <SentenceContext.Consumer>
    {({
      sentence, active, toggleActive, config,
    }) => (
      <Text sentence={sentence} active={active} toggleActive={toggleActive} config={config} />
    )}
  </SentenceContext.Consumer>
);

export default TextWithContext;
