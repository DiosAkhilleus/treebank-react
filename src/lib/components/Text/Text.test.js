import React from 'react';
import renderer from 'react-test-renderer';

import config from '../Treebank/config/arethusa-configs/aldt2grc.json';

import Text from './Text';

it('renders a sentence', () => {
  const component = (
    <Text
      sentence={{ word: [{ $: { id: '1', form: 'Ἡροδότου', postag: 'n-s---mg-' } }] }}
      active={null}
      toggleActive={() => {}}
      config={config}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('orders the words of a sentence correctly', () => {
  const sentence = {
    word: [
      { $: { id: '4', form: 'ὦ', postag: 'e--------' } },
      { $: { id: '3', form: 'χαῖρε', postag: 'v2spma---' } },
      {
        $: {
          id: '2', insertion_id: '0004f', form: '.', postag: 'u--------',
        },
      },
      {
        $: {
          id: '1', insertion_id: '0004e', form: 'κόσμε', postag: 'n-s---mv-',
        },
      },
    ],
  };

  const component = (
    <Text
      sentence={sentence}
      active={null}
      toggleActive={() => {}}
      config={config}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('styles `artificial="elliptic"` words differently', () => {
  const sentence = {
    word: [
      {
        $: {
          id: '1', form: 'Ἡροδότου', postag: 'n-s---mg-', artificial: 'elliptic',
        },
      },
    ],
  };

  const component = (
    <Text
      sentence={sentence}
      active={null}
      toggleActive={() => {}}
      config={config}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
