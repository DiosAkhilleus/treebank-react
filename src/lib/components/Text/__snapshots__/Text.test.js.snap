// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orders the words of a sentence correctly 1`] = `
<div
  className="text"
>
  <p>
    <span
      className="word"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "color": "red",
        }
      }
      tabIndex="0"
    >
      χαῖρε
    </span>
     
    <span
      className="word"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "color": "gold",
        }
      }
      tabIndex="0"
    >
      ὦ
    </span>
     
    <span
      className="word"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "color": "rgb(43, 114, 124)",
        }
      }
      tabIndex="0"
    >
      κόσμε
    </span>
     
    <span
      className="word"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "color": null,
        }
      }
      tabIndex="0"
    >
      .
    </span>
     
  </p>
</div>
`;

exports[`renders a sentence 1`] = `
<div
  className="text"
>
  <p>
    <span
      className="word"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "color": "rgb(43, 114, 124)",
        }
      }
      tabIndex="0"
    >
      Ἡροδότου
    </span>
     
  </p>
</div>
`;

exports[`styles \`artificial="elliptic"\` words differently 1`] = `
<div
  className="text"
>
  <p>
    <span
      className="word elliptic"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "color": "rgb(43, 114, 124)",
        }
      }
      tabIndex="0"
    >
      Ἡροδότου
    </span>
     
  </p>
</div>
`;
